mlflow/pyspark/ml/__init__.py

```diff
@@ -0,0 +1,302 @@
+    }
+    # TODO:
+    #  handle special case: CrossValidator.estimatorParamMaps
+
+    expanded_param_map = {}
+    for k, v in param_map.items():
+        if isinstance(v, Params):
+            # handle the case param value type inherits `pyspark.ml.param.Params`
+            # e.g. param like `OneVsRest.classifier`/`CrossValidator.estimator`
+            expanded_param_map[k] = v.uid
+            internal_param_map = _get_instance_param_map(v)
+            for ik, iv in internal_param_map.items():
+                expanded_param_map[f"{v.uid}.{ik}"] = iv
+        else:
+            expanded_param_map[k] = v
+
+    return expanded_param_map
+
+
+# NOTE: The current implementation doesn't guarantee thread-safety, but that's okay for now because:
+# 1. We don't currently have any use cases for allow_children=True.
+# 2. The list append & pop operations are thread-safe, so we will always clear the session stack
+#    once all _SparkTrainingSessions exit.
+class _SparkTrainingSession(object):
+    _session_stack = []
+
+    def __init__(self, clazz, allow_children=True):
+        self.allow_children = allow_children
+        self.clazz = clazz
+        self._parent = None
+
+    def __enter__(self):
+        if len(_SparkTrainingSession._session_stack) > 0:
+            self._parent = _SparkTrainingSession._session_stack[-1]
+            self.allow_children = (
+                _SparkTrainingSession._session_stack[-1].allow_children and self.allow_children
+            )
+        _SparkTrainingSession._session_stack.append(self)
+        return self
+
+    def __exit__(self, tp, val, traceback):
+        _SparkTrainingSession._session_stack.pop()
+
+    def should_log(self):
+        """
+        Returns True when at least one of the following conditions satisfies:
+
+        1. This session is the root session.
+        2. The parent session allows autologging and its class differs from this session's class.
+        """
+        return (self._parent is None) or (
+            self._parent.allow_children and self._parent.clazz != self.clazz
+        )
+
+
+def _get_warning_msg_for_fit_call_with_a_list_of_params(estimator):
+    return (
+        "Skip pyspark ML autologging when calling "
+        + f"{_get_fully_qualified_class_name(estimator)}.fit with a list of params,"
+        + "if you want to autolog for this case, you convert code to call `fit` with "
+        + "each single param map."
+    )
+
+
+@experimental
+@autologging_integration(AUTOLOGGING_INTEGRATION_NAME)
+def autolog(
+    log_models=True,
+    disable=False,
+    exclusive=False,
+    disable_for_unsupported_versions=False,
+    silent=False,
+):  # pylint: disable=unused-argument
+    """
+    Enables (or disables) and configures autologging for pyspark ml estimators.
+    This method is not threadsafe.
+    This API requires Spark 3.0 or above.
+
+    **When is autologging performed?**
+      Autologging is performed when you call ``Estimator.fit``
```

Comment:
```suggestion
      Autologging is performed when you call ``Estimator.fit``, except for estimators (featurizers) under pyspark.ml.feature.
```

Could we add this for now & update to indicate that featurizers will be logged as part of Pipelines when we add pipeline support?